arch/arm/soc/nordic_nrf5/nrf52/soc.c

```diff
@@ -373,6 +375,34 @@ static void clock_init(void)
 	SystemCoreClock = __SYSTEM_CLOCK_64M;
 }
 
+static void block_forever_if_built_for_wrong_chip(void)
+{
+    u32_t soc =
+        IS_ENABLED(CONFIG_SOC_NRF52832) ? 0x52832UL :
+        IS_ENABLED(CONFIG_SOC_NRF52840) ? 0x52840UL :
+        0xb0e;
+
+    if(soc == 0xb0e)
+    {
+        // We are unable to guess what SoC we have built this FW for,
+        // so we skip this test.
+        return;
+    }
+
+    if(NRF_FICR->INFO.PART != soc)
+    {
+        // Wrong BOARD.
+        //
+        // The FW has been built for an SoC that does not match this
+        // chip, to prevent hard-to-debug undefined behaviour we block
+        // forever. A block like this is easy to debug, it will show
+        // up in the stack trace when the FW is debugged.
+
+    wrong_board_block_forever:
+        goto wrong_board_block_forever;
```

Comment:
call k_panic() instead

Reply:
I assumed this code would be run before k_panic() could be run, will double-check that assumption.

Reply:
Fixed.

k_panic uses "kernel services" like the uart driver to be able to do things like printk(),
system_init is run before these kernel services are ready (PRE_KERNEL_1), so k_panic
was misbehaving (no printk's) when it was invoked here.

So I moved it to a later point (POST_KERNEL) in the boot process and created a file soc_post_kernel.c in case any other Soc-code needs kernel services in the future.

This reorganization unfortunately increased the code size from 16 bytes to 40,
but I guess we can live with that.